src/index/sparse/sparse_inverted_index.h

```diff
@@ -0,0 +1,449 @@
+    }
+
+    Status
+    Add(const void* csr_matrix, bool sealed = true) {
+        size_t rows, cols, nnz;
+        const IndPtrT* indptr;
+        const IndicesT* indices;
+        const T* data;
+        parse_csr_matrix(csr_matrix, rows, cols, nnz, indptr, indices, data);
+
+        const int32_t scale_factor = sealed ? 1 : 2;
+
+        for (size_t i = 0; i < rows + 1; ++i) {
+            indptr_.push_back(nnz_ + indptr[i]);
+        }
+
+        indices_.reserve(nnz_ + nnz);
+        indices_.insert(indices_.end(), indices, indices + nnz);
+        data_.reserve(nnz_ + nnz);
+        data_.insert(data_.end(), data, data + nnz);
+
+        if (n_cols_ < cols) {
+            n_cols_ = cols;
+            inverted_lut_.resize(n_cols_);
+            if (use_wand_) {
+                max_in_dim_.resize(n_cols_);
+            }
+        }
+
+        for (size_t i = n_rows_; i < n_rows_ + rows; ++i) {
+            for (IndPtrT j = indptr_[i]; j < indptr_[i + 1]; ++j) {
+                inverted_lut_[indices_[j]].emplace_back(i, data_[j]);
+                if (use_wand_) {
+                    max_in_dim_[indices_[j]] = std::max(max_in_dim_[indices_[j]], data_[j]);
+                }
+            }
+        }
+
+        n_rows_ += rows;
+        nnz_ += nnz;
+        return Status::success;
+    }
+
+    void
+    Search(const void* query_csr_matrix, int64_t q_id, size_t k, float drop_ratio_search, float* distances,
+           label_t* labels, size_t refine_factor, const BitsetView& bitset) const {
+        size_t len;
+        const IndicesT* indices;
+        const T* data;
+        get_row<T>(query_csr_matrix, q_id, len, indices, data);
+
+        // initially set result distances to NaN and labels to -1
+        std::fill(distances, distances + k, std::numeric_limits<float>::quiet_NaN());
+        std::fill(labels, labels + k, -1);
+        if (len == 0) {
+            return;
+        }
+
+        // if no data was dropped during both build and search, no refinement is needed.
+        if (drop_ratio_build_ == 0 && drop_ratio_search == 0) {
+            refine_factor = 1;
+        }
+
+        std::vector<std::pair<IndicesT, T>> q_vec(len);
+        for (size_t i = 0; i < len; ++i) {
+            q_vec[i] = std::make_pair(indices[i], data[i]);
+        }
+        std::sort(q_vec.begin(), q_vec.end(),
+                  [](const auto& lhs, const auto& rhs) { return std::abs(lhs.second) > std::abs(rhs.second); });
+        while (q_vec.size() && q_vec[0].second * drop_ratio_search > q_vec.back().second) {
+            q_vec.pop_back();
+        }
+
+        MinMaxHeap<T> heap(k * refine_factor);
+        if (!use_wand_) {
+            search_brute_force(q_vec, heap, bitset);
+        } else {
+            search_wand(q_vec, heap, bitset);
+        }
+
+        // no refinement needed
+        if (refine_factor == 1) {
+            collect_result(heap, distances, labels);
+        } else {
+            std::unordered_map<IndicesT, T> q_map;
```

Comment:
the size of the `q_map` is known beforehand (it is `len`). So, please construct `q_map` with the provided number of buckets (I'd try `4 * len` as a start), so 
```
std::unordered_map<IndicesT, T> q_map(4 * len);
```
The number of buckets impacts the performance of the following `dot_product()`

Reply:
Or I can tweak the performance as soon as this commit lands, there are many things to play with there :)